code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,167 @@
+#define BASE_BIOMASS_CAPACITY 1500
+#define BASE_COMPOST_CAPACITY 1500
+#define SOIL_COST 25
+#define DECAY 0.2
+#define MIN_CONVERSION 10
+/**
+  * # compost bin
+  * used to make soil from plants.
+  * Doesn't have components.
+  */
+/obj/machinery/compost_bin
+	name = "compost bin"
+	desc = "A wooden bin for composting."
+	icon = 'icons/obj/hydroponics/equipment.dmi'
+	icon_state = "compost_bin-empty"
+	power_state = NO_POWER_USE
+	density = TRUE
+	anchored = TRUE
+	/// stage in the process, for timing purposes.
+	var/process_counter = 0
+	/// amount of biomass in the compost bin
+	var/biomass = 0
+	/// amount of compost in the compost bin
+	var/compost = 0
+	/// The maximum amount of biomass the compost bin can store.
+	var/biomass_capacity = BASE_BIOMASS_CAPACITY
+	/// The maximum amount of compost the compost bin can store.
+	var/compost_capacity = BASE_COMPOST_CAPACITY
+
+/obj/machinery/compost_bin/Initialize(mapload)
+	return ..()
+
+/obj/machinery/compost_bin/on_deconstruction()
+	// returns wood instead of the non-existent components
+	new /obj/item/stack/sheet/wood(loc, 10)
+	return ..()
+
+/obj/machinery/compost_bin/screwdriver_act(mob/living/user, obj/item/I)
+	// there are no screws either
+	to_chat(user, "<span class='warning'>[src] has no screws!</span>")
+	return TRUE
+
+/obj/machinery/compost_bin/crowbar_act(mob/living/user, obj/item/I)
+	// no panel either
+	return default_deconstruction_crowbar(user, I, ignore_panel = TRUE)
+
+// Accepts inserted plants and converts them to biomass
+/obj/machinery/compost_bin/proc/make_biomass(obj/item/reagent_containers/food/snacks/grown/O)
+	// calculate biomass from plant nutriment and plant matter
+	var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+	biomass += clamp(plant_biomass * 10, 1, biomass_capacity-biomass)
+	//plant delenda est
+	qdel(O)
+
+// takes care of plant insertion and conversion to biomass, and start composting what was inserted
+/obj/machinery/compost_bin/attackby(obj/item/O, mob/user, params)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+			PB.remove_from_storage(G, src)
+			make_biomass(G)
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
```

Comment:
```suggestion
		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
			if(biomass >= biomass_capacity)
				to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
				break
			PB.remove_from_storage(G, src)
			make_biomass(G)

		if(biomass < biomass_capacity)
			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
```
Could be moved here since its checking for the same thing.

Reply:
If the bag is emptied and fills the compost bin precisely the message won't get triggered since the check in the loop needs to happen before the next item is moved from the bag so the last item will break from the loop before the check comes again. I'll try and resolve the redundancy some other way.

Reply:
I ended up moving the check to the end of the loop since I already perform it once for a different message before the loop starts.